crates/typst/src/math/stretch.rs

```diff
@@ -248,8 +248,22 @@ fn assemble(
                     .min(next.start_connector_length)
                     .scaled(ctx, base.font_size);
 
+                // Using max_overlap even when it is less than min_overlap may
+                // be considered incorrect behaviour according to the spec, but
+                // some fonts have glyphs where this is the case and they are
+                // handled just fine by other typesetting engines.
                 advance -= max_overlap;
-                growable += max_overlap - min_overlap;
+                // If max_overlap is less than min_overlap then we don't allow
+                // any growth.
+                growable += (max_overlap - min_overlap).max(Abs::zero());
+            } else {
+                // Final part's advance may not be the actual size of the part.
```

Comment:
That sounds like a font bug or intentional behavior. TeX engines notably don’t use part full advance at all and always use the glyph advance, so I don’t why you selectively do this for the last part only, either always use part full advance or always discard it.

Reply:
The problem was that the following code
```
#show math.equation: set text(font: "Concrete Math")
$ sqrt((sum_(j = 1)^d)^2) $
```
in Typst had the sqrt symbol and rule misaligned. It turns out that the bounding box for the sqrt symbol was incorrect: its height was slightly smaller than the assembled glyph. I couldn't reproduce this in LuaTeX though, and `lua-visual-debug` appeared to show the bounding box being correct.

Looking into the font, the part `/sqrt.tp` has full advance 605, but actual height 645. So we end up with the full height of the assembly being slightly shorter than it actually is, leading to the original problem. Again, I figured this is probably a bug because other engines are handling it correctly. I thought this could then only be a problem with the first and last parts in the assembly, so switching selectively for the last part seemed to make sense, matching LuaTex's behaviour whilst not changing the behaviour with other fonts.

Thanks for mentioning that LuaTeX and XeTeX don't use the part full advance at all (I couldn't tell when I looked at the code myself). Is there a specific reason the glyph advance is always used instead? (I don't quite get then why fullAdvance is even needed, when the glyph's advance would suffice...)

Reply:
> Looking into the font, the part `/sqrt.tp` has full advance 605, but actual height 645. So we end up with the full height of the assembly being slightly shorter than it actually is, leading to the original problem. Again, I figured this is probably a bug because other engines are handling it correctly. I thought this could then only be a problem with the first and last parts in the assembly, so switching selectively for the last part seemed to make sense, matching LuaTex's behaviour whilst not changing the behaviour with other fonts.

So that is indeed a font bug and the authors likely tested only in *TeX and didn’t catch it because part full advance is not used there. If it is up to me, I’d report it as a font bug and call it day. The font is likely broken in other implementations (e.g. Microsoft’s). Ignoring full advance only for the last part seems to be very arbitrary and inconsistent. Note that LuaTeX and XeTeX behavior here is peculiar as well, as they use glyph advance (at least in horizontal direction) while they should be using bounding box width, so glyphs with zero width (e.g. combining marks often get zero width automatically by font authoring tools) result in broken layout and one has to always make sure glyphs used in horizontal asssemblies have and advance width.

> Thanks for mentioning that LuaTeX and XeTeX don't use the part full advance at all (I couldn't tell when I looked at the code myself). Is there a specific reason the glyph advance is always used instead? (I don't quite get then why fullAdvance is even needed, when the glyph's advance would suffice...)

Most likely an optimization, as noted above the bounding box not the glyph advance should be used and calculating the bounding box is not cheap (especially for CFF fonts that don’t encode it at all in font data). It would allow MATH layout to be done completely without having to consult other font tables or data.

Reply:
I don't know if there is a repository for Concrete Math, but the author is listed here https://ctan.org/pkg/concmath-otf?lang=en . Hopefully he's receptive to feedback (the font seems actively maintained).

Reply:
Thank you for explaining that. I'll revert the changes here then and not ignore the full advance for only the last part. Given as well that calculating the bounding box is not cheap, I figure then it doesn't make much sense to switch to it over the full advance which is already in the GlyphPart record.

> I don't know if there is a repository for Concrete Math, but the author is listed here https://ctan.org/pkg/concmath-otf?lang=en . Hopefully he's receptive to feedback (the font seems actively maintained).

I've sent them an email. When I hear back I'll post an update.